index.js

```diff
@@ -47,49 +49,84 @@ class Rewt {
     this.options = options;
     this._redisConn = options.redisConn;
     this._scripty = new Scripty(this._redisConn);
+    this._shouldCacheSecret = options.cacheSecret || false;
   }
 
   /**
    * Signs the given payload.
    *
    * @param {Object} payload The payload to sign.
-   * @param {Function} cb Node style callback.
    */
-  sign(payload, cb) {
-    this._getSecret((err, val) => {
-      if (err) {
-        cb(err);
-      } else {
-        jwt.sign(payload, val, {}, cb);
-      }
+  async sign(payload) {
+    const secret = await this._getSecret();
+    return new Promise((resolve, reject) => {
+      jwt.sign(payload, secret, {}, (err, val) => {
+        if (err) {
+          this._clearCachedSecret();
+          return void reject(err);
+        }
+        return void resolve(val);
+      });
     });
   }
 
   /**
    * Verifies the given token and parses the payload from it.
    *
    * @param {String} token The token to verify and parse.
-   * @param {Function} cb Node style callback.
    */
-  verify(token, cb) {
-    this._getSecret((err, val) => {
-      if (err) {
-        cb(err);
-      } else {
-        jwt.verify(token, val, cb);
-      }
+  async verify(token) {
+    const secret = await this._getSecret();
+    return new Promise((resolve, reject) => {
+      jwt.verify(token, secret, (err, val) => {
+        if (err) {
+          this._clearCachedSecret();
+          return void reject(err);
+        }
+        return void resolve(val);
+      });
     });
   }
 
   /**
    * Returns the secret to sign and verify requests with.
-   * @param{Function} cb Node style callback.
+   * @returns {String} The secret to use.
+   * @throws {Error} If the secret can not be identified.
    */
-  _getSecret(cb) {
-    this._scripty.loadScript('secretRetrievalScript', secretRetrievalScript, (err, script) => {
-      if (err) return void cb(err);
+  async _getSecret() {
+    if (this._cachedSecret) {
+      // Short circuit.
+      return this._getCachedSecret();
+    }
+
+    return new Promise((resolve, reject) => {
+      this._scripty.loadScript('secretRetrievalScript', secretRetrievalScript, (err, script) => {
+        if (err) {
+          return void reject(err);
+        }
 
-      script.run(1, this._generateKeyName(), uuid.v4(), this.options.ttl, cb);
+        script.run(
+          1,
+          this._generateKeyName(),
+          uuid.v4(),
+          this.options.ttl,
+          (err, [ttl, secret]) => {
```

Comment:
Can the script return non-array type values?

Reply:
No, it'll always return an array as all return cases in the script return a Lua list.